baseapp/abci.go

```diff
@@ -1179,11 +1179,8 @@ func (app *BaseApp) handleQueryGRPC(handler GRPCQueryHandler, req *abci.RequestQ
 		return sdkerrors.QueryResult(err, app.trace)
 	}
 
-	// handler recursively can re-enter Query which requires us to give up the lock now.
-	app.mtx.Unlock()
+	// handler recursively can re-enter Query which is fine since we can acquire a read lock multiple times.
```

Comment:
I don't think we can do this. https://pkg.go.dev/sync#RWMutex

> If any goroutine calls Lock while the lock is already held by one or more readers, concurrent calls to RLock will block until the writer has acquired (and released) the lock, to ensure that the lock eventually becomes available to the writer. Note that this prohibits recursive read-locking.

Reply:
Your right. Will have to add back the unlock/lock to allow for it handled recursively.